### PR TITLE
feat: Add Create Request Form Web Component

### PR DESCRIPTION
**Pull request by sstefdev**
## Description

This pull request introduces a new Web Component for the Create Request Form, allowing users to integrate Request Network functionality into their web applications seamlessly. The component is built using Svelte and compiled to a Web Component, making it compatible across various web development environments.

## Changes

- **Web Component Integration**: Added a new compiled version of the Create Request Form as a Web Component.
- **Documentation**: Updated README.md to include installation and usage instructions for various environments including Svelte, React, and vanilla JavaScript projects.

## Motivation

The addition of the Create Request Form Web Component simplifies the integration of Request Network's functionalities into any web page or application, expanding the toolkit's versatility and accessibility. By supporting a wide range of development environments, this component ensures that developers can easily implement Request Network features regardless of the framework or library they use.

## Additional Notes

- Future updates might include more detailed customization options for the component to provide developers with enhanced control over styling and behavior.

### CODE REVIEW

**Comment by MantisClone** [comment]
At today's Daily Stand-up, we discussed which payment types the *first iteration* of the creation form should support. I think it should support:

* Native
* ERC-20
* ERC-20 Conversion

Then, in future iterations, we could add declarative, Swap-to-pay, Swap-to-conversion.

File: packages/create-request-form/package.json [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    "dist",
+    "!dist/**/*.test.*",
+    "!dist/**/*.spec.*"
+  ],
+  "peerDependencies": {
+    "svelte": "^4.0.0"
+  },
+  "devDependencies": {
+    "@requestnetwork/request-client.js": "^0.45.0",
+    "@requestnetwork/web3-signature": "^0.5.0",
```
Comment (MantisClone) [inline packages/create-request-form/package.json:35]: We should pin the request network dependencies. We strive to respect the semantic versioning rules, but I know there have been times when we didn't.

File: packages/create-request-form/src/lib/create-request-form.svelte [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+      conic-gradient(from -45deg at bottom,#0000,#000 .5deg 90deg,#0000 90.5deg) 0 100%;
+    -webkit-mask-size:25% 50%;
+    -webkit-mask-repeat:repeat-x;
+    background: linear-gradient(#25b09b 0 0) left/0% 100% no-repeat #ddd;
+    animation: l13 2s infinite linear;
+  }
+  @keyframes l13 {
+    100% {background-size: 100% 100%}
+  }
+</style>
```
Comment (MantisClone) [inline packages/create-request-form/src/lib/create-request-form.svelte:126]: Please consider adding new lines at the end of the files so Github doesn't show the red :no_entry: symbol.

File: packages/create-request-form/src/lib/create-request-form.svelte [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+    }
+
+    const requestCreateParameters = {
+      requestInfo: {
+        currency: {
+          type: Types.RequestLogic.CURRENCY.ETH,
+          value: 'ETH',
+          network: 'sepolia'
+        },
+        expectedAmount: ethers.utils.parseUnits(amount, 'ether').toString(),
```
Comment (MantisClone) [inline packages/create-request-form/src/lib/create-request-form.svelte:28]: Ethers v5 has a special function for Ether parsing.

```suggestion
        expectedAmount: ethers.utils.parseEther(amount).toString(),
```

File: packages/create-request-form/src/lib/create-request-form.svelte [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+  <input type="text" bind:value={payeeAddress} placeholder="Payee Address" />
+  <input type="text" bind:value={paymentReason} placeholder="Payment Reason" />
+  <input type="date" bind:value={dueDate} placeholder="Due Date" />
+  <button type="submit">Submit Invoice</button>
+  {#if requestInitialized}
+    <div class="confirmation-wrapper">
+      <p>Waiting for confirmation...</p>
+      <div class="loader"></div>
+    </div>
+  {/if}
```
Comment (MantisClone) [inline packages/create-request-form/src/lib/create-request-form.svelte:87]: Neat! I'd love to see how it renders.

File: packages/create-request-form/src/lib/create-request-form.svelte [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+    width: 124px;
+    height: 24px;
+    -webkit-mask: 
+      conic-gradient(from 135deg at top   ,#0000,#000 .5deg 90deg,#0000 90.5deg) 0 0   ,
+      conic-gradient(from -45deg at bottom,#0000,#000 .5deg 90deg,#0000 90.5deg) 0 100%;
+    -webkit-mask-size:25% 50%;
+    -webkit-mask-repeat:repeat-x;
+    background: linear-gradient(#25b09b 0 0) left/0% 100% no-repeat #ddd;
+    animation: l13 2s infinite linear;
+  }
```
Comment (MantisClone) [inline packages/create-request-form/src/lib/create-request-form.svelte:122]: Do you use a tool to make your styling or do you write it by hand?

File: packages/create-request-form/src/routes/+page.svelte [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+  
+  onMount(() => {
+    initializeRequestNetwork();
+  });
+</script>
+
+{#if $requestNetworkStore}
+  <CreateInvoiceForm requestNetwork={$requestNetworkStore.requestNetwork} signer={$requestNetworkStore.signer} />
+{:else}
+  <p>Loading Request Network...</p>
```
Comment (MantisClone) [inline packages/create-request-form/src/routes/+page.svelte:14]: Does initialization take a long time? About how long?

File: packages/create-request-form/src/lib/create-request-form.svelte [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+          value: payerAddress,
+        },
+        timestamp: Utils.getCurrentTimestampInSecond(),
+      },
+      paymentNetwork: {
+        id: Types.Extension.PAYMENT_NETWORK_ID.ETH_FEE_PROXY_CONTRACT,
+        parameters: {
+          paymentAddress: payeeAddress,
+          feeAddress:  ethers.constants.AddressZero,
+          feeAmount: ethers.utils.parseUnits('0.1', 'ether').toString(),
```
Comment (MantisClone) [inline packages/create-request-form/src/lib/create-request-form.svelte:44]:
```suggestion
          feeAmount: ethers.utils.parseEther('0.1').toString(),
```

File: packages/create-request-form/src/stores/requestNetworkStore.ts [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+export const requestNetworkStore: Writable<{
+  requestNetwork: RequestNetwork;
+  signer: string;
+} | null> = writable(null);
+
+export async function initializeRequestNetwork(): Promise<void> {
+  try {
+    const walletClient = createWalletClient({
+      chain: sepolia,
+      transport: custom(window.ethereum!),
```
Comment (MantisClone) [inline packages/create-request-form/src/stores/requestNetworkStore.ts:16]: So, as written, it's only compatible with injected wallets?  I'd like to support other wallets too, eventually.

File: packages/create-request-form/src/lib/create-request-form.svelte (hunk 5 above)
Comment (sstefdev) [reply packages/create-request-form/src/lib/create-request-form.svelte:122]: I wrote it by hand, just to have something fast, will add tailwind, to have faster styling

File: packages/create-request-form/src/routes/+page.svelte (hunk 6 above)
Comment (sstefdev) [reply packages/create-request-form/src/routes/+page.svelte:14]: It's practically instant, and it asks user to sign the request so we can go on with the process, of course we can change this to a button and when user click that we ask for the sign, but at the end of the day it's on the builder how he will init the request network, this was only for testing purposes to see how we would pass the RN instance to the web component

File: packages/create-request-form/src/lib/create-request-form.svelte (hunk 5 above)
Comment (MantisClone) [reply packages/create-request-form/src/lib/create-request-form.svelte:122]: There was some contention when I discussed Tailwind with @skiv71 a couple months ago. I'll schedule a brief discussion to gain alignment.

**Comment by MantisClone** [comment]
@sstefdev I want the Create Request Form use rnf_invoice v0.0.3 from the data-format package.

File: .gitignore [hunk 9]
```diff
@@ -1 +1,2 @@
 /node_modules
+.DS_Store
```
Comment (MantisClone) [inline .gitignore:2]: Please add new line at end of file

File: packages/create-request-form/vite.wc.config.ts [hunk 10]
```diff
@@ -7,7 +7,6 @@ export default defineConfig({
   plugins: [
     nodePolyfills(),
     svelte({
-      include: ["./src/lib/*.svelte"],
```
Comment (MantisClone) [inline packages/create-request-form/vite.wc.config.ts]: This is what fixed the mysterious "vite: Parse error"

Comment (MantisClone) [inline packages/create-request-form/vite.wc.config.ts]: This is what fixed the mysterious "vite Parse error".

File: packages/create-request-form/package.json [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+    "dist",
+    "!dist/**/*.test.*",
+    "!dist/**/*.spec.*"
+  ],
+  "peerDependencies": {
+    "svelte": "^4.0.0"
+  },
+  "devDependencies": {
+    "@requestnetwork/request-client.js": "0.45.0",
+    "@requestnetwork/web3-signature": "0.5.0",
```
Comment (MantisClone) [inline packages/create-request-form/package.json:35]: Should this be a peer dependency?

File: packages/create-request-form/src/app.html
Comment (MantisClone) [inline packages/create-request-form/src/app.html:1]: This is used by the regular vite build but not the web-component build.

File: packages/create-request-form/src/lib/create-request-form.svelte [hunk 12]
```diff
@@ -0,0 +1,137 @@
+<svelte:options customElement="create-request-form" />
+
+<script lang="ts">
+  import {
+    currencies,
+    getInitialFormData,
+    prepareRequestParams,
+    calculateInvoiceTotals,
+    config as generalConfig,
```
Comment (MantisClone) [inline packages/create-request-form/src/lib/create-request-form.svelte:9]:
```suggestion
    config as defaultConfig,
```

File: packages/create-request-form/src/types/types.d.ts [hunk 13]
```diff
@@ -0,0 +1,42 @@
+interface InvoiceItem {
+  description: string;
+  quantity: number;
+  unitPrice: number;
+  discount: number;
+  tax: number;
+}
+
+interface CustomFormData {
```
Comment (MantisClone) [inline packages/create-request-form/src/types/types.d.ts:9]: Please use the types defined in `data-format` package

File: packages/create-request-form/src/utils/debounce.ts
Comment (MantisClone) [inline packages/create-request-form/src/utils/debounce.ts:1]: Why does this exist?

File: packages/create-request-form/src/utils/formatAddress.ts [hunk 14]
```diff
@@ -0,0 +1,11 @@
+export const formatAddress = (
+  address: string,
+  first: number = 6,
+  last: number = 4
+): string => {
+  if (address?.length < first + last) {
+    throw new Error("Address too short to format properly.");
+  }
```
Comment (MantisClone) [inline packages/create-request-form/src/utils/formatAddress.ts:8]: This logic seems odd. Why would address length be compared with first/last name?

Comment (MantisClone) [reply packages/create-request-form/src/utils/formatAddress.ts:8]: Oh okay, it's not first/last name. It's first and last chars. Please change var name.

File: packages/create-request-form/src/utils/formatDate.ts
Comment (MantisClone) [inline packages/create-request-form/src/utils/formatDate.ts:1]: Usefule because the date picker needs a machine-readable date format, but the request view wants to show the human-readable date format.

File: packages/create-request-form/svelte.config.js [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+  kit: {
+    // adapter-auto only supports some environments, see https://kit.svelte.dev/docs/adapter-auto for a list.
+    // If your environment is not supported or you settled on a specific environment, switch out the adapter.
+    // See https://kit.svelte.dev/docs/adapters for more information about adapters.
+    adapter: adapter(),
+    alias: {
+      $src: "./src",
+      $utils: "./src/utils",
+      $stores: "./src/stores",
+      $components: "./src/components",
```
Comment (MantisClone) [inline packages/create-request-form/svelte.config.js:18]: Please remove unused aliases

File: packages/create-request-form/vite.wc.config.ts [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+    rollupOptions: {
+      plugins: [nodePolyfills({ include: ["crypto", "http"] })],
+    },
+    commonjsOptions: {
+      transformMixedEsModules: true,
+    },
+  },
+  optimizeDeps: {
+    exclude: ["@ethersproject/hash", "wrtc", "http"],
+    include: ["js-sha3", "@ethersproject/bignumber"],
```
Comment (MantisClone) [inline packages/create-request-form/vite.wc.config.ts:44]: Why are these excluded and included?

File: packages/shared/README.md
Comment (MantisClone) [inline packages/shared/README.md:1]: Please update this to be for the shared UI components, not the create-reqeust-form.

File: packages/shared/src/ui/loader.svelte
Comment (MantisClone) [inline packages/shared/src/ui/loader.svelte:1]: If this is no longer used, please remove it.

File: packages/shared/src/ui/skeleton.svelte
Comment (MantisClone) [inline packages/shared/src/ui/skeleton.svelte:1]: Add config object to make color configurable.

File: packages/shared/vite.config.ts [hunk 17]
```diff
@@ -0,0 +1,7 @@
+import { sveltekit } from "@sveltejs/kit/vite";
+import { defineConfig } from "vite";
+import { nodePolyfills } from "vite-plugin-node-polyfills";
+
+export default defineConfig({
+  plugins: [nodePolyfills(), sveltekit()],
```
Comment (MantisClone) [inline packages/shared/vite.config.ts:6]: Does shared actually need nodePolyfills?

File: packages/shared/svelte.config.js
Comment (MantisClone) [inline packages/shared/svelte.config.js:1]: Is this file required?

File: packages/create-request-form/src/lib/invoice/form.svelte [hunk 18]
```diff
@@ -0,0 +1,358 @@
...
+        type="text"
+        value={creatorId}
+        label="Creator Identification"
+        placeholder="Connect wallet to populate"
+      />
+      <Input
+        id="payeeAddress"
+        type="text"
+        value={formData.payeeAddress}
+        placeholder="From (Required)"
```
Comment (MantisClone) [inline packages/create-request-form/src/lib/invoice/form.svelte:107]: Please use the word 'address' to make it clear this is an ethereum address.
Also consider adding  "0x..." to make it more clear.

File: packages/create-request-form/src/lib/invoice/form.svelte [hunk 19]
```diff
@@ -0,0 +1,358 @@
...
+            type="text"
+            value={formData.sellerInfo?.businessName}
+            placeholder="Company Name"
+            handleInput={handleAdditionalInfo}
+          />
+          <Input
+            id="sellerInfo-taxRegistration"
+            type="text"
+            value={formData.sellerInfo?.taxRegistration}
+            placeholder="Tax Registration"
```
Comment (MantisClone) [inline packages/create-request-form/src/lib/invoice/form.svelte:137]: Please make this "Tax Identification Number (TIN)"

Comment (MantisClone) [reply packages/create-request-form/src/lib/invoice/form.svelte:137]: Do the same for buyer info.

Comment (MantisClone) [reply packages/create-request-form/src/lib/invoice/form.svelte:137]: Fix it for the view as well.

File: packages/create-request-form/src/lib/invoice/form.svelte (hunk 18 above)
Comment (MantisClone) [reply packages/create-request-form/src/lib/invoice/form.svelte:107]: Please fix it for the view as well.

File: packages/create-request-form/src/lib/invoice/form.svelte [hunk 20]
```diff
@@ -0,0 +1,358 @@
...
+      id="dueDate"
+      type="date"
+      min={inputDateFormat(formData.issuedOn)}
+      value={inputDateFormat(formData.dueDate)}
+      className="w-full"
+      label="Due Date"
+      {handleInput}
+    />
+    <Dropdown
+      placeholder="Select a currency"
```
Comment (MantisClone) [inline packages/create-request-form/src/lib/invoice/form.svelte:231]: I think we need a chain picker along with a currency picker.

File: packages/create-request-form/src/lib/invoice/form.svelte [hunk 21]
```diff
@@ -0,0 +1,358 @@
...
+    </div>
+    <div class="flex flex-col gap-[9px] ml-auto w-fit">
+      <p>Issued on {formData.issuedOn && formatDate(formData.issuedOn)}</p>
+      <p>
+        Payment due by {formData.dueDate && formatDate(formData.dueDate)}
+      </p>
+    </div>
+  </div>
+  <div class="flex gap-[20px]">
+    <div class="flex flex-col w-full max-w-[500px] gap-[20px]">
```
Comment (MantisClone) [inline packages/create-request-form/src/lib/invoice/form.svelte:94]: Please rearrange the form sections to match [this Figma](https://www.figma.com/file/cEkRPhER2ArCnOwNVDmVVA/Request?type=design&node-id=164-299&mode=design&t=d5G7FimIxA9YQ59j-0)

**Review by MantisClone — APPROVED** [review]
I'm pre-emptively approving with the expectation that the comments are addressed.

File: packages/create-request-form/src/types/types.d.ts
Comment (MantisClone) [inline packages/create-request-form/src/types/types.d.ts:1]: Please use the types in `data-format` package.  Except for `IConfig`. That one is okay to stay.

File: packages/create-request-form/src/utils/invoiceTotals.ts [hunk 22]
```diff
@@ -0,0 +1,34 @@
+export const calculateInvoiceTotals = (data: InvoiceItem[]) => {
```
Comment (MantisClone) [inline packages/create-request-form/src/utils/invoiceTotals.ts:1]: consider using getInvoiceTotal from `data-format` package.
https://github.com/RequestNetwork/requestNetwork/blob/master/packages/data-format/src/format/rnf_invoice/utils.ts#L4

File: packages/create-request-form/src/utils/invoiceTotals.ts [hunk 23]
```diff
@@ -0,0 +1,34 @@
...
+
+  return {
+    amountWithoutTax,
+    totalDiscount,
+    totalTaxAmount,
+    totalAmount,
+  };
+};
+
+export const calculateItemTotal = (item: InvoiceItem): number => {
```
Comment (MantisClone) [inline packages/create-request-form/src/utils/invoiceTotals.ts:28]: Consider using `getInvoiceLineTotal` from `data-format`
https://github.com/RequestNetwork/requestNetwork/blob/master/packages/data-format/src/format/rnf_invoice/utils.ts#L11

**Review by MantisClone — APPROVED** [review]

File: packages/create-request-form/package.json [hunk 24]
```diff
@@ -0,0 +1,69 @@
...
+  ],
+  "peerDependencies": {
+    "svelte": "^4.0.0"
+  },
+  "devDependencies": {
+    "@sveltejs/adapter-auto": "^2.0.0",
+    "@sveltejs/kit": "^1.27.4",
+    "@sveltejs/package": "^2.0.0",
+    "@sveltejs/vite-plugin-svelte": "^2.5.2",
+    "assert": "^2.1.0",
```
Comment (MantisClone) [inline packages/create-request-form/package.json:38]: Try without assert. Please remove if not needed.

File: packages/create-request-form/package.json [hunk 25]
```diff
@@ -0,0 +1,69 @@
...
+    "autoprefixer": "^10.4.19",
+    "bootstrap": "^5.3.2",
```
Comment (MantisClone) [inline packages/create-request-form/package.json:40]: Is bootstrap used? If not, please remove.

File: packages/create-request-form/package.json [hunk 26]
```diff
@@ -0,0 +1,69 @@
...
+    "browserify-zlib": "^0.2.0",
+    "crypto-browserify": "^3.12.0",
```
Comment (MantisClone) [inline packages/create-request-form/package.json:42]: Please move `browserify-zlib` and `crypto-browserify` to `dependencies`

Same for `stream-browserify`

File: packages/create-request-form/package.json [hunk 27]
```diff
@@ -0,0 +1,69 @@
...
+    "ethers": "^5.7.2",
```
Comment (MantisClone) [inline packages/create-request-form/package.json:43]: Please remove `ethers` it's not used.

File: packages/create-request-form/package.json [hunk 28]
```diff
@@ -0,0 +1,69 @@
...
+    "postcss": "^8.4.38",
+    "publint": "^0.2.7",
+    "rollup-plugin-polyfill-node": "^0.13.0",
+    "stream-browserify": "^3.0.0",
+    "svelte": "^4.0.5",
+    "svelte-check": "^3.6.0",
+    "tailwindcss": "^3.4.3",
+    "tslib": "^2.4.1",
+    "typescript": "^5.0.0",
+    "viem": "^2.9.15",
```
Comment (aimensahnoun) [inline packages/create-request-form/package.json:53]: Viem needs to be a regular dependancy

File: packages/shared/svelte.config.js
Comment (MantisClone) [reply packages/shared/svelte.config.js:1]: required.

File: packages/shared/vite.wc.config.ts
[the remark this answers is not in the document]
Comment (MantisClone) [reply packages/shared/vite.wc.config.ts:1]: required.

File: packages/create-request-form/src/types/types.d.ts
Comment (MantisClone) [reply packages/create-request-form/src/types/types.d.ts:1]: We decided to keep the custom types for now.

File: packages/create-request-form/src/utils/invoiceTotals.ts (hunk 22 above)
Comment (MantisClone) [reply packages/create-request-form/src/utils/invoiceTotals.ts:1]: Decided to use custom utils for now.

File: packages/create-request-form/src/utils/invoiceTotals.ts (hunk 23 above)
Comment (MantisClone) [reply packages/create-request-form/src/utils/invoiceTotals.ts:28]: Decided to use custom utils for now.

**Review by MantisClone — APPROVED** [review]
